File: internal/util/events.go
package util

import "github.com/wailsapp/wails/v3/pkg/application"

// Event name constants. Every custom Wails event used by the app is named here.
// Events whose payload types live in this package are registered in init below;
// events with payload types in other packages (config, notes, kernel_manager,
// jupyter_protocol/sockets) are registered in those packages' init functions.
// Each event must be registered exactly once — application.RegisterEvent panics
// on duplicates — and registration gives emit-time payload validation plus
// generated TypeScript typings for the frontend.
const (
	// File events
	EventFileCreate = "file:create"
	EventFileDelete = "file:delete"
	EventFileRename = "file:rename"
	EventFileWrite  = "file:write"

	// Folder events
	EventFolderRename = "folder:rename"
	EventFolderDelete = "folder:delete"
	EventFolderCreate = "folder:create"

	// UI events
	EventZoomIn              = "zoom:in"
	EventZoomOut             = "zoom:out"
	EventZoomReset           = "zoom:reset"
	EventSettingsOpen        = "settings:open"
	EventSearchNote          = "search:note"
	EventFullscreen          = "window:fullscreen"
	EventToggleSidebar       = "sidebar:toggle"
	EventSidebarFilesOpen    = "sidebar:files:open"
	EventFileTreeFilterFocus = "file-tree:filter:focus"
	EventFileTreeContentDrop = "file-tree:content-drop"
	EventEditorContentDrop   = "editor:content-drop"

	// File watcher events
	EventSettingsUpdate    = "settings:update"
	EventTagsUpdate        = "tags:update"
	EventTagsIndexUpdate   = "tags:index_update"
	EventSavedSearchUpdate = "saved-search:update"
	EventCodeResultsUpdate = "code-results:update"

	// Kernel instance events (per-instance, not per-language)
	EventKernelInstanceCreated     = "kernel:instance:created"
	EventKernelInstanceShutdown    = "kernel:instance:shutdown"
	EventKernelInstanceStatus      = "kernel:instance:status"
	EventKernelInstanceHeartbeat   = "kernel:instance:heartbeat"
	EventKernelInstanceLaunchError = "kernel:instance:launch_error"
	EventKernelInstanceExited      = "kernel:instance:exited"

	// Code block events (scoped by messageId)
	EventCodeBlockStream        = "code:code-block:stream"
	EventCodeBlockExecuteResult = "code:code-block:execute_result"
	EventCodeBlockDisplayData   = "code:code-block:display_data"
	EventCodeBlockExecuteInput  = "code:code-block:execute_input"
	EventCodeBlockStatus        = "code:code-block:status"
	EventCodeBlockIopubError    = "code:code-block:iopub_error"
	EventCodeBlockInputRequest  = "code:code-block:input_request"
	EventCodeBlockInspectReply  = "code:code-block:inspect_reply"
	EventCodeBlockCompleteReply = "code:code-block:complete_reply"
	EventCodeBlockExecuteReply  = "code:code-block:execute_reply"
)

// A map of folderAndNoteNames to tags
type TagsUpdateEventData map[string][]string

// FolderCreateEventData represents the data structure for folder create events
type FolderCreateEventData struct {
	FolderPath string `json:"folderPath"`
}

// FolderDeleteEventData represents the data structure for folder delete events
type FolderDeleteEventData struct {
	FolderPath string `json:"folderPath"`
}

// FolderRenameEventData represents the data structure for folder rename events
type FolderRenameEventData struct {
	OldFolderPath string `json:"oldFolderPath"`
	NewFolderPath string `json:"newFolderPath"`
}

// FileCreateEventData represents the data structure for file create events
type FileCreateEventData struct {
	FilePath string `json:"filePath"`
}

// FileDeleteEventData represents the data structure for file delete events
type FileDeleteEventData struct {
	FilePath string `json:"filePath"`
}

// FileRenameEventData represents the data structure for file rename events
type FileRenameEventData struct {
	OldFilePath string `json:"oldFilePath"`
	NewFilePath string `json:"newFilePath"`
}

// FileWriteEventData represents the data structure for file write events
type FileWriteEventData struct {
	FilePath string `json:"filePath"`
	Markdown string `json:"markdown,omitempty"`
}

// ContentDropEventData represents dropped OS files over a registered drop target
// (file tree or editor).
type ContentDropEventData struct {
	DroppedFiles    []string `json:"droppedFiles"`
	TargetElementID string   `json:"targetElementId,omitempty"`
	X               int      `json:"x"`
	Y               int      `json:"y"`
}

func init() {
	// File and folder events are emitted as batches accumulated per debounce
	// cycle by the file watcher, hence the slice payloads.
	application.RegisterEvent[[]FileCreateEventData](EventFileCreate)
	application.RegisterEvent[[]FileDeleteEventData](EventFileDelete)
	application.RegisterEvent[[]FileRenameEventData](EventFileRename)
	application.RegisterEvent[[]FileWriteEventData](EventFileWrite)
	application.RegisterEvent[[]FolderCreateEventData](EventFolderCreate)
	application.RegisterEvent[[]FolderDeleteEventData](EventFolderDelete)
	application.RegisterEvent[[]FolderRenameEventData](EventFolderRename)

	application.RegisterEvent[application.Void](EventZoomIn)
	application.RegisterEvent[application.Void](EventZoomOut)
	application.RegisterEvent[application.Void](EventZoomReset)
	application.RegisterEvent[application.Void](EventSettingsOpen)
	application.RegisterEvent[application.Void](EventSearchNote)
	application.RegisterEvent[bool](EventFullscreen)
	application.RegisterEvent[application.Void](EventToggleSidebar)
	application.RegisterEvent[application.Void](EventSidebarFilesOpen)
	application.RegisterEvent[application.Void](EventFileTreeFilterFocus)
	application.RegisterEvent[ContentDropEventData](EventFileTreeContentDrop)
	application.RegisterEvent[ContentDropEventData](EventEditorContentDrop)

	application.RegisterEvent[TagsUpdateEventData](EventTagsUpdate)
	application.RegisterEvent[application.Void](EventTagsIndexUpdate)
	application.RegisterEvent[application.Void](EventSavedSearchUpdate)
}

File: frontend/src/utils/events.ts
import { Window } from '@wailsio/runtime';
import { WailsEvent } from '../hooks/events';

// Event name constants — mirrors internal/util/events.go
// Frontend-only events are noted below.

// File events
export const FILE_CREATE = 'file:create';
export const FILE_DELETE = 'file:delete';
export const FILE_RENAME = 'file:rename';
export const FILE_WRITE = 'file:write';

// Folder events
export const FOLDER_RENAME = 'folder:rename';
export const FOLDER_DELETE = 'folder:delete';
export const FOLDER_CREATE = 'folder:create';

// UI events
export const ZOOM_IN = 'zoom:in';
export const ZOOM_OUT = 'zoom:out';
export const ZOOM_RESET = 'zoom:reset';
export const SETTINGS_OPEN = 'settings:open';
export const SEARCH_NOTE = 'search:note';
export const FULLSCREEN = 'window:fullscreen';
export const WINDOW_RELOAD = 'window:reload';
export const TOGGLE_SIDEBAR = 'sidebar:toggle';
export const SIDEBAR_FILES_OPEN = 'sidebar:files:open';
export const SIDEBAR_SEARCH_OPEN = 'sidebar:search:open';
export const FILE_TREE_CONTENT_DROP = 'file-tree:content-drop';
export const EDITOR_CONTENT_DROP = 'editor:content-drop';

// Context Menu events

// File watcher events
export const SETTINGS_UPDATE = 'settings:update';
export const TAGS_INDEX_UPDATE = 'tags:index_update';
export const SAVED_SEARCH_UPDATE = 'saved-search:update';

// Kernel instance events (per-instance, not per-language)
export const KERNEL_INSTANCE_CREATED = 'kernel:instance:created';
export const KERNEL_INSTANCE_SHUTDOWN = 'kernel:instance:shutdown';
export const KERNEL_INSTANCE_STATUS = 'kernel:instance:status';
export const KERNEL_INSTANCE_HEARTBEAT = 'kernel:instance:heartbeat';
export const KERNEL_INSTANCE_LAUNCH_ERROR = 'kernel:instance:launch_error';
export const KERNEL_INSTANCE_EXITED = 'kernel:instance:exited';

// Code block events (still scoped by messageId)
export const CODE_BLOCK_STREAM = 'code:code-block:stream';
export const CODE_BLOCK_EXECUTE_RESULT = 'code:code-block:execute_result';
export const CODE_BLOCK_DISPLAY_DATA = 'code:code-block:display_data';
export const CODE_BLOCK_EXECUTE_INPUT = 'code:code-block:execute_input';
export const CODE_BLOCK_STATUS = 'code:code-block:status';
export const CODE_BLOCK_IOPUB_ERROR = 'code:code-block:iopub_error';
export const CODE_BLOCK_INPUT_REQUEST = 'code:code-block:input_request';

// Frontend-only events (not in events.go)
export const CODE_BLOCK_INSPECT_REPLY = 'code:code-block:inspect_reply';

export async function isEventInCurrentWindow(data: WailsEvent) {
  const windowName = await Window.Name();
  if (windowName !== data.sender) return false;
  return true;
}
